File: Simulate_waveforms.cpp
#include <cstddef>
#include <iostream>
#include <vector>

#include "TMath.h"
#include "TRandom3.h"
#include "TTree.h"
#include "TFile.h"

using namespace std;
// --- Configuration ---
const double SAMPLING_PERIOD = 8.0;   // 8 ns
const int N_SAMPLES = 32;             // 32 points
const double ANALOG_STEP = 0.01;      // 10 ps resolution for simulation
const double NOISE_LEVEL = 0.05;      // Noise amplitude
const int N_EVENTS = 1;            // Number of events

// Bi-Exponential Pulse Function
double GetPulseValue(double t, double t0, double amp, double noise) {
	if (t < t0) {return noise;}
	double dt = t - t0;
	double tau_rise = 1.5;
	double tau_fall = 45.0;
	// Normalized approx
	return amp * (TMath::Exp(-dt / tau_fall) - TMath::Exp(-dt / tau_rise)) + noise;
}

void Simulate_waveforms() {
	TFile *f = new TFile("SiPM_Simulation.root", "RECREATE");

	// ======================================================
	// TREE 1: Configuration (The Constants)
	// ======================================================
	// We create a tree just to store the setup parameters
	TTree *tConfig = new TTree("Config", "Run Parameters");

	// Variables to hold the constants
	double cfg_sampling_period = SAMPLING_PERIOD;
	double cfg_analog_step_ns = ANALOG_STEP;


	// Create Branches
	tConfig->Branch("sampling_period_ns", &cfg_sampling_period, "sampling_period_ns/D");
	tConfig->Branch("analog_step_ns", &cfg_analog_step_ns, "analog_step_ns/D");

	// FILL ONCE: We only need to save this info one time
	tConfig->Fill(); 


	// ======================================================
	// TREE 2: The Data (The Waveforms)
	// ======================================================

	cout << "Filling config tree: " << endl;

	TTree *tWave = new TTree("Waveforms", "Simulated SiPM Data");

	unsigned int n_events_simulated = N_EVENTS;

	int event_id;
	double true_time;
	std::vector<double>* waveform = new vector<double>();
	std::vector<double>* true_waveform = new vector<double>();

	// Note: We removed the 'time_axis' array from here because 
	// we can now reconstruct it using the 'Config' tree!

	tWave->Branch("event_id", &event_id, "event_id/I");
	tWave->Branch("true_time", &true_time, "true_time/D");
	tWave->Branch("digitized_waveform", waveform, "waveform");
	tWave->Branch("true_waveform", true_waveform, "true_waveform");

	cout << "Config tree filled " << endl;

	TRandom3 *rnd = new TRandom3(0);

	cout << "Filling tree " << endl;
	for (int i = 0; i < n_events_simulated; i++) {
		event_id = i;
		true_time = 20.0 + rnd->Gaus(0, 2);  // true time in nanoseconds

		// True waveform
		for (int s = 0; s < N_SAMPLES * (SAMPLING_PERIOD / ANALOG_STEP); s++){
			double t = s * ANALOG_STEP;
			double true_signal = GetPulseValue(t, true_time, 1.0, 0.0);
			true_waveform->push_back(true_signal);
		}

		
		for (int s = 0; s < N_SAMPLES; s++) {
			// Reconstruct time using the constant we know
			double t = s * SAMPLING_PERIOD; 
			double noise = rnd->Gaus(0, NOISE_LEVEL);
			double signal = GetPulseValue(t, true_time, 1.0, noise);
			waveform->push_back(signal);
		}
		tWave->Fill();
		waveform->clear();
		true_waveform->clear();
	}


	// ======================================================
	// SAVE EVERYTHING
	// ======================================================
	// This writes both Trees to the file
	f->Write(); 
	f->Close();

	std::cout << "Done! File contains 'Waveforms' and 'Config' trees." << std::endl;

}

File: Simulate_waveforms.py
import numpy as np
import ROOT
from array import array
import sys

import config as cf

# t_axis -- array of double, t0 -- detection time, double; tau_rise/fall -- rise/fall time of waveform, double; amplitude -- maximum value, double
def generate_sipm_pulse(t_axis, t0, tau_rise = 8.0, tau_fall = 170.0, amplitude = 100.0):
    """Generates a normalized bi-exponential pulse starting at t0."""
    mask = (t_axis >= t0)
    dt = t_axis[mask] - t0
    pulse = np.zeros_like(t_axis)
    pulse[mask] = amplitude * (np.exp(-dt / tau_fall) - np.exp(-dt / tau_rise))
    return pulse

# --- Simulation Parameters ---

# Setup Time Axes
total_time = cf.N_DIGITAL_POINTS * (1 / cf.DIGITAL_FS) # ~256 ns
t_analog = np.arange(0, total_time, 1/cf.ANALOG_FS)
t_digital = np.arange(0, total_time, 1/cf.DIGITAL_FS)

# --- ROOT File Setup ---
output_file = ROOT.TFile("SiPM_Simulation.root", "RECREATE")
tree = ROOT.TTree("Waveforms", "Simulated SiPM Data")

# --- Define Branches ---
# 1. Single values (Event ID, True Time)
event_id = array('i', [0])
true_time = array('d', [0.0])

# 2. Arrays (The Waveform)
# ROOT needs a fixed size array for simple branches, or std::vector for dynamic
# Here we use a C-style array of size 32
waveform_samples = array('d', [0.0]*cf.N_DIGITAL_POINTS)
true_waveform = np.zeros(len(t_analog), dtype=np.float64)

tree.Branch("event_id", event_id, "event_id/I")
tree.Branch("true_time", true_time, "true_time/D")
tree.Branch("waveform", waveform_samples, f"waveform[{cf.N_DIGITAL_POINTS}]/D")
tree.Branch("true_waveform", true_waveform, f"true_waveform[{len(t_analog)}]/D")


print(f"Generating {cf.N_EVENTS} events...")
np.set_printoptions(threshold=sys.maxsize)

# --- Simulation Loop ---
for i in range(cf.N_EVENTS):
    event_id[0] = i
    
    # 1. Randomize Arrival Time (gaussian around 20 ns)
    t0 = 40 + np.random.normal(0.0, 5.0)
    true_time[0] = t0

    # 2. Generate and save Analog Pulse

    amplitude = cf.AMPLITUDE_VALUE + np.random.uniform(-300, 300)
    analog_waveform = generate_sipm_pulse(t_analog, t0=t0, amplitude=amplitude)
    np.copyto(true_waveform, analog_waveform)

    # Check true waveform contents
    # print("Saving true_waveform to .txt file")
    # np.savetxt("plots/true_waveform.txt", true_waveform);
    # print("printing analog_waveform to .txt file")
    # np.savetxt("plots/analog_waveform.txt", analog_waveform);

    # 3. Add Noise
    noise = np.random.normal(0, 2.5, len(analog_waveform))
    analog_waveform += noise

    # 4. Digitize (Interpolate to 8ns grid)
    digitized = np.interp(t_digital, t_analog, analog_waveform)

    # 5. Fill ROOT Branch
    for j in range(cf.N_DIGITAL_POINTS):
        waveform_samples[j] = digitized[j]

    tree.Fill()

# --- Write & Close ---
tree.Write()
output_file.Close()

print("Done! Data saved to 'SiPM_Simulation.root'")
